### PR TITLE
feat(cfg): control-flow graph engine, formal-verification stack, and CFG-backed rules

### PR DESCRIPTION
**Pull request by aidenybai**
## Summary

What started as a CFG beef-up grew into a dedicated control-flow analysis engine. This PR extracts the per-function control-flow graph into a new internal **`@react-doctor/cfg`** package, layers a small formal-verification stack on top of it (SSA, dataflow, typestate, path feasibility), and lands 6 new CFG-backed rules plus several false-positive fixes on existing ones. The package is pure-TS, bundled into `oxlint-plugin-react-doctor` at build time (published surface unchanged), and lazy — a rule that never reads a layer pays nothing.

### `@react-doctor/cfg` — the engine

- **Structural CFG.** Each basic block is a typed instruction list ending in a first-class `Terminal` modeled on the React Compiler HIR taxonomy (`goto` / `if` / `switch` / loops / `logical` / `ternary` / `optional` / `try` / `return` / `throw`), with `fallthrough` join blocks and explicit `goto` lowering of `break` / `continue`.
- **Expression-level control flow.** Ternary arms, `&&` / `||` / `??` (and `&&=` / `||=` / `??=`) right operands, and optional-chain links past each `?.` each get their own blocks — so a hook / `setState` short-circuited inside an expression is correctly seen as conditional.
- **Analyses.** Reachability, dominance (Cooper–Harvey–Kennedy idom tree over RPO), post-dominance, dominance frontier (Cytron), loop membership, unreachable blocks, infinite-loop folding (`isInfiniteLoopStart`, oxc-parity constant folding), `try`/`catch`/`finally` finalize/join semantics, and a Graphviz `toDot` export.
- **Native SSA.** Variable-level static single assignment via the Braun et al. (2013) on-the-fly sealed-block algorithm (the same one the React Compiler's `EnterSSA` uses) plus redundant-φ elimination — a clean-room port (MIT attribution, no Babel). Query API: `versionAt`, `reachingDefinition`, `isLiveValue`, `isRedefinedBetween`, `bindingOf`.
- **Formal-verification layers** (all pure-TS, lazy, run once per scan):
  - **Dataflow** — `solveDataflow`, a generic monotone worklist fixpoint over a `Lattice<Fact>`, with `analyzeDefiniteAssignment` built on it.
  - **Typestate** — `verifyTypestate(cfg, { automaton, classifier })`, a reusable resource-protocol automaton reporting illegal transitions and leaked resources.
  - **Path feasibility** — a bounded, dependency-free checker (`isPathFeasible` + `lowerGuard` / `pathConditionFacts`) that refutes correlated-branch counterexamples via union-find congruence closure. It only ever *suppresses* a diagnostic when the path search is complete and every counterexample is provably infeasible, so it strictly removes false positives and is never unsound for bug-finding.
- **Parity corpora ported as tests** from oxc (`no-unreachable`, `no-fallthrough`, `no-unsafe-finally`, `getter-return`), ESLint code-path analysis, and React Compiler `BuildHIR`.

### New rules (CFG-backed)

- **`correctness/no-unreachable-code`** (Bugs) — code that never runs because every path above returns / throws / breaks / continues / loops forever. ESLint `no-unreachable` carve-outs (hoisted functions, type-only TS decls, bare `var x;`).
- **`correctness/no-dead-assignment`** (SSA) — a write to a reassignable local whose value is never read because every path overwrites it first. Quiet for `const`, compound assignments, and closure-captured bindings.
- **`correctness/no-use-before-define`** (dataflow) — a block-scoped binding used lexically before its declaration in the same synchronous execution (a TDZ `ReferenceError`). Sound by construction: quiet for hoisted `var`/functions, params, globals, and closure/class-body access.
- **`state-and-effects/effect-cleanup-not-on-every-path`** — a subscription/timer acquired in an effect whose returned cleanup is skipped on an early-return path. Complements `effect-needs-cleanup` (which only checks a cleanup exists at all).
- **`state-and-effects/no-unreleased-resource`** (typestate) — a resource opened in an effect callback and released *inline* on some paths but leaked on an early return. Scoped to `useEffect`/`useLayoutEffect`/`useInsertionEffect`; the returned-cleanup contract stays owned by `effect-cleanup-not-on-every-path`.
- **`state-and-effects/no-stale-closure-capture`** — a `useMemo`/`useCallback` closure capturing a `let` reassigned later in the same render. Quiet for `const`, never-reassigned bindings, and the deferred effect hooks.
- **`state-and-effects/no-set-state-in-render-loop`** (Bugs) — a `useState` setter called inside a render-phase loop. Partitions cleanly with `no-set-state-in-render` on `isUnconditionalFromEntry` (no double-reporting).

### Existing-rule fixes (consumer-visible)

- **`nextjs-no-redirect-in-try-catch`** — stop mis-flagging `redirect()` / `notFound()` in a `catch`, in a `finally`, or in a `try` whose only companion is `finally` (none swallow the navigation control-flow error). Replaced the try-depth counter with an ancestry walk.
- **`no-set-state-in-render`** — now flags any setter the CFG proves runs on every render path (`const x = setCount(...)`, unconditional blocks), not just bare top-level statements; the guarded store-previous-render pattern stays quiet.
- **`no-mutating-reducer-state`** — drop remembered mutations that can't reach the same-reference return (e.g. `for (…) { state.items.push(x); return { ...state } }` with a trailing `return state` on the no-match path).
- **`js-hoist-regexp` / `js-index-maps` / `js-set-map-lookups`** — loop-awareness keys off real CFG loop membership instead of lexical nesting depth, so work inside a callback that merely escapes a loop is no longer flagged.

### Investigated, deliberately not shipped

- **`server-auth-actions`** — `isUnconditionalFromEntry` gating produced ERROR-severity false positives on idiomatic validation-before-auth.
- **`effect-needs-cleanup`** — CFG reachability regressed real patterns where the resource is created in a `forEach`/helper closure (cross-function). The positional heuristic is better there.

## Test plan

- [x] `pnpm test` (full monorepo) — all packages green, including the new `@react-doctor/cfg` suites and ported parity corpora
- [x] `pnpm typecheck` clean
- [x] `pnpm lint` clean
- [x] `pnpm format:check` clean
- [x] New rule suites + the rule-metadata convention guards (title length / dash-free copy / category bucketing)
- [ ] RDE parity vs OSS corpus

---

> [!NOTE]
> **Medium Risk**
> Large new analysis engine and several new global/path-sensitive rules can shift diagnostic volume; behavior is heavily tested via parity corpora and differential tests, but edge cases in CFG/SSA may still surface in the wild.
> 
> **Overview**
> Introduces internal **`@react-doctor/cfg`** (bundled into the plugin at build time): a per-function structural CFG with React Compiler–style terminals, expression-level lowering (`&&` / ternary / optional chaining), dominance/reachability/loop primitives, Braun SSA, a generic dataflow solver, typestate verification, and bounded path-feasibility pruning. CFG/SSA/definite-assignment are shared per file via a `Program`-keyed cache with lazy derived analyses (~2× faster for multi-rule scans).
> 
> **New rules** use these layers: `no-unreachable-code`, `no-dead-assignment`, `no-use-before-define`, `effect-cleanup-not-on-every-path`, `no-unreleased-resource`, `no-stale-closure-capture`, `no-set-state-in-render-loop`.
> 
> **Existing rules** are tightened or de-noised: CFG-backed `no-set-state-in-render` and loop rules; redirect-in-try-catch, reducer mutation, and loop-aware JS rules lose false positives; CFG fixes include correct `break` inside `switch`-in-loop and try-body liveness for dead-assignment.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit df5a163c70efa98c3703159a7da679b8e30b4b5d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-no-redirect-in-try-catch.ts [hunk 1]
```diff
@@ -2,8 +2,27 @@ import { NEXTJS_NAVIGATION_FUNCTIONS } from "../../constants/nextjs.js";
...
+  let child: EsTreeNode = callExpression;
+  let ancestor = callExpression.parent ?? null;
+  while (ancestor) {
+    if (isNodeOfType(ancestor, "TryStatement") && ancestor.handler && ancestor.block === child) {
+      return true;
+    }
+    child = ancestor;
+    ancestor = ancestor.parent ?? null;
+  }
+  return false;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-no-redirect-in-try-catch.ts:23]: Disagreeing with this one — it's a false alarm. A `redirect()`/`notFound()` that throws inside a nested `finally` (or a finally-only inner `try`) propagates out of that try via finally-completion, so if it lexically sits in an **outer** `try` body that has a `catch`, the outer catch genuinely swallows it. That's a true positive, exactly what the rule targets (cf. the existing `flags an inner-catch redirect that the outer try body swallows` case).

The walk only matches when it steps up *through a try block* (`ancestor.block === child`), never through a `catch`/`finally` clause, so the precision cases still stay quiet — e.g. a redirect in a nested `finally` with no enclosing `catch` is not reported.

Locked all three in as regression tests in 7e40c19c:
- nested `finally` inside an outer try-with-catch → flagged
- finally-only inner `try` inside an outer try-with-catch → flagged
- nested `finally` with no catch anywhere → quiet

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-use-before-define.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        const declarationStart = nodeStart(symbol.declarationNode);
+        if (accessStart < 0 || declarationStart < 0) return;
+        if (accessStart >= declarationStart) return;
+
+        if (crossesDeferredBoundary(reference.scope, symbol.scope)) return;
+
+        context.report({
+          node,
+          message: `"${symbol.name}" is used here before its declaration runs, which throws a ReferenceError from the Temporal Dead Zone. Move this access below the declaration.`,
+        });
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-use-before-define.ts:126]: This one is working as intended, so I'm resolving it rather than changing the rule.

`typeof` is exempt from the TDZ `ReferenceError` only for *never-declared* identifiers; for a `let`/`const` that is in its Temporal Dead Zone, `typeof x` still throws:

```js
(() => { typeof x; let x = 1; })();
// ReferenceError: Cannot access 'x' before initialization
typeof neverDeclared; // "undefined"  (no throw)
```

The rule only flags references that resolve to a `let`/`const`/`class`/`using` binding, so `typeof` of an undeclared name is never flagged, while `typeof` of a binding in its TDZ is — matching the runtime. Both behaviors are locked with tests in d1b6230c.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-use-before-define.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        if (!reference) return;
+
+        const symbol = reference.resolvedSymbol;
+        if (!symbol) return;
+        if (!TDZ_BINDING_KINDS.has(symbol.kind)) return;
+
+        const accessStart = nodeStart(node);
+        const declarationStart = nodeStart(symbol.declarationNode);
+        if (accessStart < 0 || declarationStart < 0) return;
+        if (accessStart >= declarationStart) return;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-use-before-define.ts]: Fixed in d1b6230c. A read inside the binding's own initializer (`const x = x`) is now flagged — the binding is uninitialized until its initializer finishes evaluating, so the read always hits the TDZ. A read in a closure inside the initializer (`const x = () => x`) is still allowed, since it runs later. Test added.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-stale-closure-capture.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+      const reportedBindings = new Set<number>();
+      for (const reference of closureCaptures(closure, context.scopes)) {
+        const symbol = reference.resolvedSymbol;
+        if (!symbol || (symbol.kind !== "let" && symbol.kind !== "var")) continue;
+        if (reportedBindings.has(symbol.id)) continue;
+
+        const binding = context.ssa.bindingOf(reference.identifier);
+        if (binding === null) continue;
+        if (!context.ssa.isRedefinedBetween(node, renderExit, binding)) continue;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-stale-closure-capture.ts]: Fixed in 7f3c8f71. Added `ssa.isRedefinedAfter(fromNode, binding)` (true when the binding is reassigned on any path reachable from the hook call, with no endpoint constraint) and switched the rule to it, so a reassignment on a branch that returns before the render's final statement is now flagged. Test added.

**Comment by rayhanadev** [comment]
## Performance: CFG analysis layer (~2× faster)

Each rule rebuilt the CFG, SSA, and typestate analysis from scratch, and a few graph passes ran super-linear. This change shares one analysis per file and drops those passes to linear time, with identical diagnostics.

**Measured** (before/after from git worktrees at the same base, re-parsing each iteration, median of 9 runs):

| workload | before | after | speed-up |
|---|--:|--:|--:|
| analysis layer (9 rules × 27 files, parse subtracted) | 162.7 ms | 83.1 ms | **1.96×** |
| full per-file pass, with parsing | 182.1 ms | 102.4 ms | 1.78× |
| `analyzeControlFlow` + queries, 400-branch function | 202.4 ms | 104.6 ms | 1.93× |

The microbench gain grows with function size (1.59× at 50 branches, 1.93× at 400), tracking the O(V³)→O(V+E) passes.

### Each change

| Speed-up | Where | Before | After | Improvement |
|---|---|---|---|---|
| Share analysis across rules | `wrap-with-semantic-context.ts` | per-rule `cachedCfg = analyzeControlFlow(root)` | module `WeakMap<Program, …>`, `shared.cfg ??= analyzeControlFlow(root)` | scopes **12×→1×**, CFG **7×→1×**, SSA **2×→1×** per file |
| Reuse one CFG in SSA and dataflow | `ssa.ts`, `definite-assignment.ts` | `analyzeControlFlow(program)` | `controlFlowArg ?? analyzeControlFlow(program)` | CFG build **2×→1×** per cfg+ssa rule |
| Build per-function analyses lazily | `control-flow-graph.ts` | 6 structures built eagerly for every function | `get dominatorTree(){ return (d ??= compute()) }` | only the queried structure is built |
| Loop membership | `loops.ts` | BFS from every block to test self-reachability | one Tarjan SCC pass | **O(V³)→O(V+E)** |
| BFS dequeue | `reachability.ts`, `unconditional.ts` | `queue.shift()` | `queue[head++]` | **O(V²)→O(V+E)** per traversal |
| `enclosingFunction` lookup | `control-flow-graph.ts` | walk the parent chain on every call | `WeakMap` memo | O(depth)→O(1) on repeat |
| Successor/predecessor arrays | `block-edges.ts` | `block.successors.map(e => e.to)` each call | `WeakMap` memo | drops a per-call allocation in the dominator and dataflow fixpoints |
| Child traversal | `for-each-child-node.ts` | `for (const k of Object.keys(node))` | `for (const k in node)` | drops a `string[]` allocation on the hottest per-node call |
| Path enumeration | `enumerate-paths.ts` | per-path length and goal-count caps | adds a `MAX_PATH_VISITS` budget | bounds the worst case when goal blocks are sparse |
| Typestate error pass | `verify.ts` | replays every block | skips when `automaton.errorStates.size === 0` | drops a full block sweep for leak-only automata |

<details><summary>Headline snippets</summary>

Share one analysis per file across all rules (`wrap-with-semantic-context.ts`):
```ts
// before: every wrapped rule rebuilds its own
let cachedCfg = null;
const getCfg = () => (cachedCfg ??= analyzeControlFlow(programRoot));

// after: one build per Program, shared across all rules
const analysisCache = new WeakMap<EsTreeNode, SharedAnalysis>();
const getCfg = () => {
  const shared = sharedAnalysisFor(programRoot);
  return (shared.cfg ??= analyzeControlFlow(programRoot));
};
```

Loop membership, O(V³) to O(V+E) (`loops.ts`):
```ts
// before: per-block self-reachability BFS
for (const start of cfg.blocks) {
  // BFS over successors; mark start cyclic if it reaches itself
}

// after: one strongly-connected-components pass
//   cyclic = blocks in an SCC of size > 1, plus self-loops
```
</details>

No behavior change: every rule produced byte-identical diagnostics before and after, across an adversarial corpus and the fixtures.
